### PR TITLE
refactor: replace is-my-json-valid with Ajv

### PR DESCRIPTION
**Pull request by gajus**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[X] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

Replace the ESLint rule validation library from is-my-json-valid to Ajv.

**Is there anything you'd like reviewers to focus on?**

WIP. Will ping when ready for review.

**Motivation**

Ajv has more contributors, more downloads, more regular releases. https://npmcompare.com/compare/ajv,is-my-json-valid summarizes it quite well.

From the technical perspective, the Ajv is the fastest and the most standard compliant (https://github.com/epoberezkin/ajv#performance) implementation of a JSON validator.

It is being used in webpack. I am using it in https://www.npmjs.com/package/table, https://www.npmjs.com/package/isomorphic-webpack, etc.

Finally, Ajv was originally planned to be used for ESLint (https://github.com/eslint/eslint/pull/8295#issuecomment-296556738). The current error message formatting is a bit hacky as a result. Ajv (optionally) provides a custom error message abstraction via https://github.com/epoberezkin/ajv-errors.

### CODE REVIEW

File: lib/util/ajv.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+//------------------------------------------------------------------------------
+
+const ajv = new Ajv({
+    meta: false,
+    validateSchema: false,
+    verbose: true
+});
+
+ajv.addMetaSchema(metaSchema);
+/* eslint no-underscore-dangle: "off" */
```
Comment (platinumazure) [inline lib/util/ajv.js]: I believe this will disable the rule for the whole file. I think you might want to use `// eslint-disable-next-line no-underscore-dangle`instead?

Comment (epoberezkin) [inline lib/util/ajv.js]: thank you, I updated

File: lib/rules/comma-dangle.js [hunk 2]
```diff
@@ -81,49 +81,49 @@ module.exports = {
...
-                            "always",
-                            "always-multiline",
-                            "only-multiline",
-                            "never"
-                        ]
-                    },
-                    valueWithIgnore: {
-                        anyOf: [
-                            {
-                                $ref: "#/defs/value"
```
Comment (not-an-aardvark) [inline lib/rules/comma-dangle.js]: I haven't looked at this thoroughly yet, but as a quick question: Would this be a breaking change for rules? If some third-party rules have been relying on the (apparently nonstandard) `$ref` property from `is-my-json-valid`, it seems like that won't work anymore if `$ref` is no longer supported in ajv.

Comment (epoberezkin) [inline lib/rules/comma-dangle.js]: @not-an-aardvark this change only fixes the schema, it is not related to ajv. Ajv fully supports $ref (and it is standard), the changed schema also has $ref's. Should it be in a separate PR maybe? @gajus?

Comment (not-an-aardvark) [inline lib/rules/comma-dangle.js]: Oh I see, I misunderstood what the change was here. I guess the more general question is: is this schema fix necessary, or would the PR still work without this schema fix? (I don't mind modifying the schema in general, but if it requires changes then that might indicate that switching to Ajv would risk breaking third-party rules.)

Comment (epoberezkin) [inline lib/rules/comma-dangle.js]: @not-an-aardvark , now I understand the problem here. When you pass array as a rule schema (`schema: [obj]`, where obj as a JSON schema) eslint internally converts it to a following JSON schema:

```javascript
{
  type: 'array',
  items: [
    obj
  ],
  minItems: 0,
  maxItems: 1
}
```

If there are definitions inside obj (like it is with comma-dangle) then their JSON pointers should be extended as well, otherwise they are not found. is-my-json-valid quietly ignores $ref's that are not found and ajv (by default) throws exception. This change to the schema fixes this issue. In addition there are "$refs" instead of "$ref" in several places, that ignored by both is-my-json-valid and by ajv.

There may be some third party rules with similar incorrect usage of definitions not on the top level but inside item schema and these rules will start logging errors. Should it be a breaking change? I am not sure. These rules are not working correctly now so maybe it is ok if they have to be fixed. There are the following options:

1. Leave things as they are (therefore breaking third-party rules with incorrect schemas)
2. Change Ajv settings to log warnings when $ref's are not found instead of throwing exception, so things will work as they do now and all rules can be fixed separately (including comma-dangle.
3. Keep things completely as they are, quietly ignoring $ref's that cannot be found.
4. Improve eslint to correctly extend JSON pointers in the above scenario, so internally "#/defs/value" would become "#/items/0/defs/value". That's my least favourite option, as it can be confusing to debug and some third-party rules can still start breaking as currently they simply don't validate against these schemas and then they start, so if something was invalid it would start breaking.

I think option 2 is the safest, but let me know what you think.

Comment (not-an-aardvark) [inline lib/rules/comma-dangle.js]: That makes sense, thanks for explaining. I agree that option 2 seems safest, if it's not too difficult a change. Maybe we can switch to option 1 in a future major release.

Comment (epoberezkin) [inline lib/rules/comma-dangle.js]: Thank you.
Option 2 is a very simple change, I will do it. And I will move fix for comma-dangle in a separate PR.
Option 4 could be a good change for the next major version as users probably expect that all items are validated separately and therefore use $ref's, expecting them to work, rather than ignored. So instead of saying "you can't use $ref if your rule schema is an array" or "you have to understand internal mechanics of eslint to use $ref in such cases" we can just make things work as users expect, either by validating items separately (as users think eslint does) or by transforming schemas correctly taking $ref's into account.

Comment (epoberezkin) [inline lib/rules/comma-dangle.js]: It is done, fix to comma-dangle and docs update on $ref usage are in #8864

**Comment by gajus** [comment]
LGTM

File: lib/config/config-validator.js [hunk 3]
```diff
@@ -181,10 +186,10 @@ function formatErrors(errors) {
  * @returns {void}
  */
 function validateConfigSchema(config, source) {
-    const validator = schemaValidator(configSchema, { verbose: true });
+    validateSchema = validateSchema || ajv.compile(configSchema);
```
Comment (zxqfox) [inline lib/config/config-validator.js]: Personally I'd put this `function` to block code with `validateSchema` variable like:
```js
{
    let validateSchema;

    /**
     * ...
     */
    function validateConfigSchema(...) {
        validateSchema = validateSchema || ajv.compile(configSchema);
        // ...
    }
}
```
It would to read this part of code as well as help v8 to reduce variables in context.

Or at least just put it closer to this function.

File: lib/util/ajv.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * @author Evgeny Poberezkin
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const Ajv = require("ajv"),
+    metaSchema = require("ajv/lib/refs/json-schema-draft-04.json");
```
Comment (zxqfox) [inline lib/util/ajv.js]: There is already draft-06, hope you know.

File: lib/config/config-validator.js (hunk 3 above)
Comment (epoberezkin) [inline lib/config/config-validator.js]: @zxqfox is it better to make a commit now or make a separate PR after this one merged?

File: lib/util/ajv.js (hunk 4 above)
Comment (epoberezkin) [inline lib/util/ajv.js]: Ajv supports draft-06 by default - this code is needed as the schemas for all rules would be draft-04 and there are changes in draft-06 that would break them. So I thought that migrating to draft-06 would require a major version change as some third party rules will need to change their schemas.

File: lib/config/config-validator.js (hunk 3 above)
Comment (zxqfox) [inline lib/config/config-validator.js]: I'm not sure we ever need that. It's just my personal preferences. Prob no need to do that at all ;-)

File: lib/util/ajv.js (hunk 4 above)
Comment (zxqfox) [inline lib/util/ajv.js]: Ah, right. Forgot about we checking options for rules with JSON schemas. Sorry and thank you.

Comment (epoberezkin) [inline lib/util/ajv.js]: Thank you!

**Comment by not-an-aardvark** [comment]
I'm not sure this actually needs to be approved by the TSC since it's an just internal change, but it seems like there was some confusion about whether it needed TSC approval, so I'll add it to the agenda just in case.

**TSC Summary**: This PR proposes switching the internal schema validator to use `ajv` rather than `is-my-json-valid`.

**TSC Question**: Should we accept this proposal?

File: lib/config/config-validator.js (hunk 3 above)
Comment (epoberezkin) [inline lib/config/config-validator.js]: There can be an eslint rule for that ;)

**Comment by not-an-aardvark** [comment]
This was accepted in [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/df5bb3c39c6a87f0e11dbb1e5c0fda7f54563972/notes/2017/2017-07-06.md#refactor-replace-is-my-json-valid-with-ajv).

**Comment by not-an-aardvark** [comment]
Thanks for contributing!

**Comment by gajus** [comment]
Thank you for accepting this promptly.
